dolphin/api/v1/alert.py

```diff
@@ -0,0 +1,153 @@
+from dolphin import exception
+from dolphin import utils
+from dolphin.api import validation
+from dolphin.api.common import wsgi
+from dolphin.api.schemas import alert as schema_alert
+from dolphin.api.views import alert as alert_view
+from dolphin.drivers import api as driver_api
+from dolphin.i18n import _
+
+LOG = log.getLogger(__name__)
+
+
+SNMPv3_keys = ('username', 'password', 'auth_enable', 'auth_protocol', 'encryption_enable',
+               'encryption_protocol', 'encryption_password', 'engine_id')
+
+
+class AlertController(wsgi.Controller):
+    def __init__(self):
+        super().__init__()
+        self.driver_api = driver_api.API()
+
+    @wsgi.response(200)
+    @validation.schema(schema_alert.put)
+    def put(self, req, id, body):
+        """Create a new alert source or update an exist one."""
+        ctx = req.environ['dolphin.context']
+        alert_source = body
+
+        try:
+            alert_source["storage_id"] = id
+            db.storage_get(ctx, id)
+            alert_source = self._input_check(alert_source)
+
+            if self._exist(ctx, id):
+                alert_source = db.alert_source_update(ctx, id, alert_source)
+            else:
+                alert_source = db.alert_source_create(ctx, alert_source)
+        except exception.StorageNotFound as e:
+            msg = (_("Alert source cannot be created or updated for a non-exist storage %s.") % id)
+            raise exc.HTTPBadRequest(explanation=msg)
+        except exception.InvalidInput as e:
+            msg = _('Failed to put alert_source: {0}'.format(e))
+            raise exc.HTTPBadRequest(explanation=msg)
+
+        return alert_view.build_alert_source(self._decrypt_password(alert_source))
+
+    @wsgi.response(200)
+    def show(self, req, id):
+        ctx = req.environ['dolphin.context']
+        try:
+            alert_source = db.alert_source_get(ctx, id)
+        except exception.AlertSourceNotFound as e:
+            raise exc.HTTPNotFound(explanation=e.msg)
+        except Exception as e:
+            LOG.error(e)
+            raise exc.HTTPBadRequest(explanation=e.msg)
+
+        return alert_view.build_alert_source(self._decrypt_password(alert_source))
+
+    @wsgi.response(200)
+    def delete(self, req, id):
+        ctx = req.environ['dolphin.context']
+        try:
+            db.alert_source_delete(ctx, id)
+        except exception.AlertSourceNotFound as e:
+            raise exc.HTTPNotFound(explanation=e.msg)
+
+    def _input_check(self, alert_source):
+        version = alert_source.get('version')
+
+        if version.lower() == 'snmpv3':
+            user_name = alert_source.get('username', None)
+            auth_enabled_str = alert_source.get('auth_enabled', None)
+            engine_id = alert_source.get('engine_id', None)
+            if not user_name or not auth_enabled_str or not engine_id:
+                msg = "If snmp version is SNMPv3, then username, auth_enabled and engine_id are required."
```

Comment:
Few lines looks to be beyond page border

Reply:
My IDE(Pycharm) shows, all lines are fine, which IDE you used?

Reply:
Anyway, I cut those long lines.

Reply:
As per Python coding guidelines, the max char for a line is 79..So maybe you can set the line wrap/length accordingly